compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGenOptimizationReportEmitter.java

```diff
@@ -0,0 +1,176 @@
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.wso2.ballerinalang.compiler.util.ProjectDirConstants.TARGET_DIR_NAME;
+
+/**
+ * Emits time durations and optimized node details for codegen optimization.
+ *
+ * @since 2201.10.0
+ */
+public class CodeGenOptimizationReportEmitter {
+
+    private static final Map<PackageID, Long> birOptimizationDurations = new LinkedHashMap<>();
+    private static long nativeOptimizationDuration = 0;
+    private static final PrintStream out = System.out;
+    private static final String CODEGEN_OPTIMIZATION_REPORT = "codegen_optimization_report.json";
+
+    protected static void flipBirOptimizationTimer(PackageID pkgId) {
+        if (birOptimizationDurations.containsKey(pkgId)) {
+            long totalDuration = System.currentTimeMillis() - birOptimizationDurations.get(pkgId);
+            birOptimizationDurations.put(pkgId, totalDuration);
+        } else {
+            birOptimizationDurations.put(pkgId, System.currentTimeMillis());
+        }
+    }
+
+    protected static void flipNativeOptimizationTimer() {
+        if (nativeOptimizationDuration == 0) {
+            nativeOptimizationDuration = System.currentTimeMillis();
+        } else {
+            nativeOptimizationDuration = System.currentTimeMillis() - nativeOptimizationDuration;
+        }
+    }
+
+    protected static void emitBirOptimizationDuration() {
+        long totalDuration = birOptimizationDurations.values().stream().mapToLong(Long::longValue).sum();
+
+        out.printf("Duration for unused BIR node analysis : %dms%n", totalDuration);
+        birOptimizationDurations.forEach((key, value) -> out.printf("    %s : %dms%n", key, value));
+        out.println();
+        // TODO make this class a singleton
+        birOptimizationDurations.clear();
+    }
+
+    protected static void emitNativeOptimizationDuration() {
+        out.println(
+                "Duration for Bytecode Optimization (analysis + deletion) : " + nativeOptimizationDuration + "ms");
+        nativeOptimizationDuration = 0;
+    }
+
+    protected static void emitOptimizedExecutableSize(Path executableFilePath) {
+        try {
+            float optimizedJarSize = Files.size(executableFilePath) / (1024f * 1024f);
+            System.out.printf("Optimized file size : %f MB%n", optimizedJarSize);
+        } catch (IOException e) {
+            throw new ProjectException("Failed to emit optimized executable size ", e);
+        }
+    }
+
+    protected static void emitCodegenOptimizationReport(
+            Map<PackageID, UsedBIRNodeAnalyzer.InvocationData> invocationDataMap, Path projectDirectoryPath,
+            ProjectKind projectKind) {
+
+        Path reportParentDirectoryPath = projectDirectoryPath.resolve(TARGET_DIR_NAME).toAbsolutePath().normalize();
+
```

Comment:
We can remove this new line.

Reply:
Fixed in 7ac189d8d68a32b7f7175045b7c8a1db5d78f3ef.